lib/pedicel/ec.rb

```diff
@@ -19,18 +24,23 @@ def decrypt(symmetric_key: nil, merchant_id: nil, certificate: nil, private_key:
       # | token's ephemeralPublicKey | | certificate   | '------------------'
       # '----------------------------' '---------------'
 
-      if private_key
-        symmetric_key = symmetric_key(private_key: private_key,
-                                      certificate: certificate,
-                                      merchant_id: merchant_id)
+      if !certificate.nil? && merchant_id.nil?
+        merchant_id = self.class.merchant_id(certificate: certificate)
+      end
+
+      if private_key && merchant_id
+        symmetric_key = symmetric_key(
+          private_key: private_key,
+          merchant_id: merchant_id
+        )
```

Comment:
Why? The initial check ("assert style-ish" programming) shouldn't be carried "all over" IMO.

`symmetric_key` can itself derive the `merchant_id`; no reason to copy the code.

Reply:
It's a bit of pedantry I guess.
When called with (`symmetric_key` and `private_key`), it would fail in `symmetric_key()`, since neither `certificate` nor `merchant_id` was defined, so I added
```ruby
+      if private_key && merchant_id
￼+        symmetric_key = symmetric_key(
￼+          private_key: private_key,
￼+          merchant_id: merchant_id
￼+        )
```

I guess the check above came as a side-effect of that.

Though, it would probably make sense to argue that the function should fail if passed only `symmetric_key` and `private_key`. which it would with the original code.

My thought when I wrote it was probably that it wasn't entirely clear.

Reply:
The initial check would fail if `symmetric_key && private_key`.

Reply:
No it doesn't. A unit test like the following fails:
```ruby
describe 'decrypt'
  context 'should fail on' do
    it 'excessive arguments' do
      backend, merchant, token, _data = PedicelPay::Helper.generate_all

      p = Pedicel::EC.new(
        JSON.parse(token.to_json),
        config: Pedicel::DEFAULTS.merge(
          trusted_ca_pem: backend.ca_certificate
        )
      )

      merchant_id = merchant.merchant_id
      certificate = merchant.certificate
      private_key = merchant.key
      symmetric_key = p.symmetric_key(
        certificate: merchant.certificate,
        private_key: merchant.key
      )
      expect do
        p.decrypt(
          symmetric_key: symmetric_key,
          private_key: private_key
        )
      end.to raise_error(ArgumentError)
    end
  end
end
```

With `symmetric_key` and `private_key` the right side evaluates to `false`, while the left side evaluates to `true`, satisfying the XOR.